fix(ivy): correct ngtsc path handling in Windows

JoostK opened the pull request:
As it turns out, the usage of `path.posix` does not unify path handling
across operating systems. Instead, `canonical-path` is used to ensure
path handling is consistent, avoiding incorrect paths in Windows.

See https://github.com/angular/angular/pull/25862#discussion_r216157914

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

See https://github.com/angular/angular/pull/25862#discussion_r216157914,  https://github.com/angular/angular/issues/25880#issuecomment-431647422 and https://github.com/angular/angular/issues/26436#issuecomment-429579282

## What is the new behavior?

Using `canonical-path` ensures correct path handling, as confirmed by yours truly and https://github.com/angular/angular/pull/25862#discussion_r216157914

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

The `canonical-path.d.ts` ambient file was moved from ngcc to the ngtsc/util target, such that it can be used by all deps. I am uncertain whether or not `"@ngdeps//canonical-path"` should be included as deps in all targets, or if the current transitive dependency is okay.

//cc @alxhub @petebacondarwin

petebacondarwin commented:
I really should just add the typings to the canonical-path project :-)

JoostK commented:
Hmm, the old build pipeline (build.sh on Travis) fails spectacularly because of that ambient file :( I'd need some help to resolve this failure.

IgorMinar reviewed (commented):
This doesn't look right to me. That canonical-path dependency is a runtime dependency, which means we'll need to add it to our [package.json](https://github.com/angular/angular/blob/master/packages/compiler-cli/package.json)..... OMG how did we end up with all of these dependencies for @angular/compiler-cli?  @petebacondarwin 👀

In general we want as few dependencies as possible because they make our packages bigger and less reliable (and additionally, 3rd party dependencies can become a problem in google3 where we have only very restricted set of dependencies available).

@alxhub can you take a look at this one as well please?

I ran a g3 presubmit and it fails because of `ParsedPath` not being a known type: https://test.corp.google.com/ui#id=OCL:218454158:BASE:218454273:1540353573542:33ba48e0

petebacondarwin commented:
The `ParsedPath` error is a bug in my typings file.

JoostK committed:
fix(ivy): correct ngtsc path handling in Windows

As it turns out, the usage of path.posix does not unify path handling
across operating systems. Instead, canonical-path is used to ensure
path handling is consistent, avoiding incorrect paths in Windows.

See https://github.com/angular/angular/pull/25862#discussion_r216157914

JoostK commented:
Rebased onto #26719, only second commit is relevant (and smaller than before)

JoostK commented:
Rebased again now that #26719 has landed, this should be good to go.

IgorMinar commented:
Caretaker note: no g3 impact.

kara commented:
@IgorMinar @alexeagle This PR had some unforeseen consequences in G3. Adding the `canonical-path` dependency breaks the build in G3 because we don't have that package in third-party. We should probably discuss if we want to sync all of canonical-path into G3 just for the path resolution before merging this in?

alexeagle commented:
Discussed with @vikerman - we don't want the compiler to take on a new third-party dependency. We have only `typescript` as a dependency and want to keep it this way. We have solved the same thing in angular-cli and rules_typescript simply by replacing backslash with forward slash in the right places - see https://github.com/bazelbuild/rules_typescript/pull/160/files for example

petebacondarwin commented:
Implementing `normalizeSlashes()` in ngtsc works for me, but we need to work out how to enforce its use. As @alexeagle points out - we need to build and test on Windows.

petebacondarwin commented:
Closing this as it is outdated, conflicted and not the approach that we wish to take. As agreed offline @JoostK or @gkalpak will implement the alternative solution, which is to implement helper functions in this project, and in the future to require that a "branded" string is required for all path related activities.
